KeenClient/KeenLogger.m

```diff
@@ -0,0 +1,136 @@
+//
+//  KeenLogger.m
+//  KeenClient
+//
+//  Created by Brian Baumhover on 2/18/17.
+//  Copyright © 2017 Keen Labs. All rights reserved.
+//
+
+#import "KeenLogger.h"
+#import "KeenLogSinkNSLog.h"
+
+@implementation KeenLogger
+
+BOOL logSinksEnabled;
+_Atomic(BOOL) loggingEnabled;
+KeenLogLevel logLevel;
+NSMutableArray *logSinks;
+dispatch_queue_t loggerQueue;
+
+
++ (KeenLogger*)sharedLogger {
+    static KeenLogger* logger;
+    
+    // This black magic ensures this block
+    // is dispatched only once over the lifetime
+    // of the program. It's nice because
+    // this works even when there's a race
+    // between threads to create the object,
+    // as both threads will wait synchronously
+    // for the block to complete.
+    static dispatch_once_t predicate = {0};
+    dispatch_once(&predicate, ^{
+        logger = [[KeenLogger alloc] init];
+    });
+    
+    return logger;
+}
+
+
+- (KeenLogger*)init {
+    self = [super init];
+    
+    if (nil != self) {
+        // By default we won't enable logging.
+        logSinksEnabled = NO;
+        loggingEnabled = NO;
+        // Default log level is error so we aren't noisy
+        logLevel = KeenLogLevelError;
+        // Create a serial queue for logging messages and manipulating loggers.
+        // This will ensure that messages from different threads
+        // don't get mixed together, and we don't have to worry
+        // about concurrent access when adding and removing loggers
+        // since those operations will be dispatched to this queue as well.
+        loggerQueue = dispatch_queue_create("io.keen.logger", DISPATCH_QUEUE_SERIAL);
+        if (NULL == loggerQueue) {
+            // Failed to allocate the queue
+            self = nil;
+        }
+        
+        logSinks = [[NSMutableArray alloc] init];
+        if (nil == logSinks) {
+            // Failed to allocate the array
+            self = nil;
+        }
+    }
+    
+    return self;
+}
+
+
+- (void)disableLogging {
+    loggingEnabled = NO;
+    dispatch_async(loggerQueue, ^() {
+        // Disable actual logging to sinks
+        // on the queue so anything that
+        // has already been logged but hasn't
+        // been sinked will have the
+        // opportunity to do so.
+        logSinksEnabled = NO;
+    });
+}
+
+
+- (void)enableLogging {
+    loggingEnabled = YES;
+    dispatch_async(loggerQueue, ^() {
+        if ([logSinks count] == 0) {
+            // Allocate a default logger if none have
+            // been provided.
+            [logSinks addObject:[[KeenLogSinkNSLog alloc] init]];
```

Comment:
Yes, that's true. I'd argue that we don't worry about it, since this is basically to preserve existing functionality when the SDK is used as before. I find it unlikely that someone will want to enable NSLog logging for a while and then disable it in favor of a different logger during an app's runtime. I suppose the fix would be to keep track of the logger that has been added implicitly and remove it when logging is disabled.

What bothers me more is the semantic coupling the SDK now has with the client, but that's a trade-off to retain existing SDK behavior. Or perhaps the way we fix that is to always enable the NSLog sink unless the client explicitly calls a new enableLoggingWithEnableNSLog:(BOOL)shouldEnableNSLog?.